---
params:
  pheno_name:
    label: Phenotype
    value: Ins_tAUC
  coefType:
    choices:
    - coef
    - blup
    input: select
    label: Coefficient calculation
    value: coef
  dataSetup:
    choices:
    - dataJax.R
    - dataMadison.R
    - dataJaxMadison.R
    input: select
    label: Data setup file
    value: dataMadison.R
  docTitle:
    label: Document title
    value: Mediation Scan
  snpScan:
    input: numeric
    label: Width for SNP scan (0 = whole chromosome)
    value: 10
  offset:
    label: Offset width for fine mapping
    value: 2
  resultpath:
    input: text
    label: Path to put CSV results (none if '')
    value: '.'
  docAuthor:
    label: Author
    value: '[Brian S Yandell](brian.yandell@wisc.edu)'
  docOutput:
    choices:
    - html_document
    - pdf_document
    - word_document
    input: select
    label: Output format
    value: html_document
  datapath:
    input: text
    label: Path to derived data (required for Madison)
    value: ..
  showPeaks:
    label: Show highest peaks (0=no, n=number to show)
    value: 0
  useIntcov:
    choices:
    - no
    - yes
    label: Use interacting covariate (sex)?
    value: no
  extraSteps:
    label: Exra steps file (optional)
    value: ''
---

---
title: "`r params$docTitle`"
author: "`r params$docAuthor`"
date: "`r  format(Sys.time(), '%d %B %Y')`"
output:
  "`r params$docOutput`":
    toc: true
    toc_depth: 3
    code_folding: hide
---

```{r setup, include=FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(ggplot2)
library(plotly)
```

This is built off script developed by Dan Gatti. In Rstudio, use `Knit with Parameters` from <kbd>Knit</kbd> pulldown to adjust parameter settings. See [README.md](README.md) for details.

```{r}
pheno_name <- params$pheno_name
datapath <- params$datapath
resultpath <- params$resultpath
```

```{r}
data.frame(params_value = unlist(params))
```

### Data Setup

```{r}
source(params$dataSetup)
```

```{r}
# If looking for peaks, show them and exit.
if(params$showPeaks > 0 & exists("peaks")) {
  source("dataPeaks.R")
  knitr::knit_exit()
}
```

```{r}
addcovar <- qtl2pattern::covar_df_mx(covar)
```

```{r}
if(!is.logical(useIntcov <- params$useIntcov)) {
  useIntcov <- (useIntcov == "yes")
}
print(useIntcov)
if(("sex" %in% names(covar)) & useIntcov) {
  intcovar <- qtl2pattern::covar_df_mx(covar[, "sex", drop = FALSE])
} else {
  intcovar <- NULL
}
```

### Genome scan.

```{r ins_scan1}
genoprobs <- query_probs()$probs
map <- query_probs()$map
cat("scan1 of whole genome\n", file = stderr())
qtl_ins = qtl2::scan1(genoprobs = genoprobs, pheno = pheno_data[,pheno_name,drop=FALSE], 
            kinship = kinship, addcovar = addcovar, intcovar = intcovar, cores = 2)
autoplot(qtl_ins, map, main = pheno_name)
```

### QTL Peaks

```{r ins_peaks}
peaks = qtl2::find_peaks(qtl_ins, map, threshold = 8)
chr_id <- peaks$chr[1]
pos_Mbp <- peaks$pos[1]
peaks
```

```{r}
autoplot(qtl_ins, map[chr_id], main = pheno_name) +
  geom_vline(xintercept = pos_Mbp, linetype = "dashed")
```

### BLUP or Coef

```{r ins_coef}
if(params$coefType == "blup") {
  cat(paste("scan1blup of chromosome", chr_id, "\n"), file = stderr())
  coef_ins = qtl2::scan1blup(genoprobs = genoprobs[,chr_id],
                   pheno = pheno_data[,pheno_name,drop=FALSE], 
                   kinship = kinship[[chr_id]], 
                   addcovar = addcovar, cores = 4)
} else {
  cat(paste("scan1coef of chromosome", chr_id, "\n"), file = stderr())
  coef_ins = qtl2::scan1coef(genoprobs = genoprobs[,chr_id],
                   pheno = pheno_data[,pheno_name,drop=FALSE], 
                   kinship = kinship[[chr_id]], 
                   addcovar = addcovar, intcovar = intcovar)
}
autoplot(coef_ins, map, scan1_output = qtl_ins, main = pheno_name)
```

### Mediation with allele probs.

```{r peak_mar,message=FALSE}
peak_mar <- qtl2::find_marker(map, chr_id, pos_Mbp)
ap <- qtl2::pull_genoprobpos(genoprobs, peak_mar)
```

```{r}
mrna <- query_mrna(chr_id)
```

```{r med_scan,fig.width=9,fig.height=6}
med_scan <- intermediate::mediation_scan(target   = pheno_data[,pheno_name,drop=FALSE],
                      mediator = mrna$expr,
                      annotation = mrna$annot,
                      covar    = addcovar,
                      intcovar = intcovar,
                      kinship = kinship[[chr_id]],
                      driver = ap,
                      method = "double-lod-diff")
ggplotly(autoplot(med_scan) +
           ggtitle(pheno_name))
```

```{r}
med_lod <- intermediate::mediator_lod(
  mediator = mrna$expr,
  driver = ap,
  annotation = mrna$annot,
  kinship = kinship[[chr_id]],
  covar_med = covar, intcovar = intcovar)
med_signif <- med_lod$lod >= 5
# Add info column.
med_lod$info <- paste("chr =", med_lod$chr)
```

```{r med_test,fig.width=9,fig.height=6}
med_test <- intermediate::mediation_test(
  target   = pheno_data[,pheno_name,drop=FALSE],
  mediator = mrna$expr[, med_signif, drop = FALSE],
  annotation = mrna$annot,
  covar_tar = covar,
  covar_med = covar,
  kinship = kinship[[chr_id]],
  intcovar = intcovar,
  driver = ap)
ggplotly(autoplot(med_test))
```

```{r}
(sum_med <- summary(med_test))
```

```{r}
if(resultpath != "")
  write.csv(sum_med, file = paste0(pheno_name, "_mediation.csv"))
```

```{r med_triad,fig.width=9,fig.height=6}
med_triad <- intermediate::mediation_triad(
  target   = pheno_data[,pheno_name,drop=FALSE],
  mediator = mrna$expr[, sum_med$id[1], drop = FALSE],
  driver = ap,
  covar_tar = covar,
  kinship = kinship[[chr_id]],
  sdp = sum(2 ^ (4:7)))
autoplot(med_triad, tname = pheno_name, mname = sum_med$symbol[1], dname = peak_mar)
```

### Association mapping

Map all of Chr `r chr_id` unless `snpScan` (`r params$snpScan`) is positive.

```{r}
snpScan <- as.numeric(params$snpScan)
if(is.na(snpScan) || snpScan <= 0) {
  start <- end <- NULL
} else {
  start <- pos_Mbp - snpScan
  end <- pos_Mbp + snpScan
}
```

```{r ins_assoc_chr,fig.width=12}
if(is.null(start) & is.null(end)) {
  cat(paste("scan1snps of chromosome", chr_id, "\n"), file = stderr())
} else {
  cat(paste("scan1snps of chromosome", chr_id, "from", start, "to", end, "\n"), file = stderr())
}
  
assoc_ins = qtl2::scan1snps(genoprobs = genoprobs[,chr_id], map = map, 
                      pheno = pheno_data[,pheno_name,drop=F], kinship = kinship[[chr_id]],
                      addcovar = addcovar, intcovar = intcovar, chr = chr_id, start = start,
                      end = end, query_func = query_variant, cores = 4,
                      keep_all_snps = TRUE)
autoplot(assoc_ins$lod, assoc_ins$snpinfo, show_all_snps = FALSE)
```

Re-scan with SNP with highest LOD score to see evidence for second peak.

```{r ins_assoc_rescan,fig.width=12}
(ts <- qtl2::top_snps(assoc_ins$lod, assoc_ins$snpinfo) %>%
  arrange(desc(lod))) %>%
  head(10)
```

```{r}
if(resultpath != "")
  write.csv(ts, file = paste0(pheno_name, "_topsnps.csv"))
```

```{r ts_kludge}
# This part is clumsy but it works.
ts1 = ts[1,]
# We need to change the index to get qtl2::genoprob_to_snpprob() to work.
ts1$index = 1

# This converts the top SNP to SNP probs using the allele probs.
sp = qtl2::genoprob_to_snpprob(genoprobs = genoprobs[,chr_id], snpinfo = ts1)[[1]]

# Just keep the A column.
m <- match(rownames(covar), rownames(sp), nomatch = 0)
covar_ts = qtl2pattern::covar_df_mx(data.frame(covar[m>0,], sp[m,1,1]))
```

```{r assoc_ins2}
assoc_ins2 = qtl2::scan1snps(genoprobs = genoprobs[,chr_id], map = map, 
                      pheno = pheno_data[,pheno_name,drop=F], kinship = kinship[[chr_id]],
                      addcovar = covar_ts, intcovar = intcovar, 
                      chr = chr_id, start = start, end = end,
                      query_func = query_variant, cores = 4,
                      keep_all_snps = TRUE)
autoplot(assoc_ins2$lod, assoc_ins2$snpinfo, show_all_snps = FALSE)
```

Look at the peak at `r ts1$pos` Mb.

```{r ins_assoc_left,fig.width=9,fig.height=6}
offset = as.numeric(params$offset)
start = ts1$pos - offset
end   = ts1$pos + offset
assoc_ins_peak = qtl2::scan1snps(genoprobs = genoprobs[,chr_id], map = map, 
                      pheno = pheno_data[,pheno_name,drop=F], kinship = kinship[[chr_id]],
                      addcovar = addcovar, intcovar = intcovar,
                      chr = chr_id, start = start, end = end,
                      query_func = query_variant, cores = 4,
                      keep_all_snps = TRUE)

genes = query_gene(chr = chr_id, start = start, end = end)
genes = genes[(genes$stop - genes$start) < 2,]
autoplot(assoc_ins_peak$lod, assoc_ins_peak$snpinfo, genes = genes, top_panel_prop = 0.5, colors = "black")
```

View patterns in top SNPs.

```{r assoc_ins_pattern,fig.width=9,fig.height=6}
autoplot(assoc_ins_peak$lod, assoc_ins_peak$snpinfo, patterns = "all", drop_hilit = 5, main = pheno_name)
```

### Top missense, splice or stop SNPs

```{r ins_top_snps_left}
ts <- qtl2::top_snps(assoc_ins_peak$lod, assoc_ins_peak$snpinfo, drop = 1) %>%
  arrange(desc(lod))
left_join(ts %>% dplyr::rename(id = "ensembl_gene"),
          mrna$annot %>% select(id, symbol),
          by = "id") %>%
  select(snp_id, lod, pos, id, symbol, consequence) %>%
  head(10)
```

```{r}
if(resultpath != "")
  write.csv(left_join(ts %>% dplyr::rename(id = "ensembl_gene"),
          mrna$annot %>% select(id, symbol),
          by = "id"), file = paste0(pheno_name, "_topsnps_peak.csv"))
```

```{r}
ts %>% dplyr::filter(grepl("splice|missense|stop", consequence))
```

### Mediation with SNP probs at ~85 Mb

```{r, ins_med_snps_peak,message=FALSE}
ts <- qtl2::top_snps(assoc_ins_peak$lod, assoc_ins_peak$snpinfo)
ts <- ts[which.max(ts$lod),]
```

```{r peak_snp,message=FALSE}
sp_peak <- qtl2::genoprob_to_snpprob(pairprobs, assoc_ins_peak$snpinfo)
peak_snp <- 
  qtl2::find_index_snp(
    assoc_ins_peak$snpinfo,
    qtl2::find_marker(
      assoc_ins_peak$snpinfo,
      ts$chr, ts$pos))
sp <- sp_peak[[chr_id]][,, peak_snp]
```

```{r med2_scan,fig.width=9,fig.height=6}
med2_scan <- intermediate::mediation_scan(target  = pheno_data[,pheno_name,drop=FALSE],
                       mediator = mrna$expr,
                       annotation = mrna$annot,
                       covar    = addcovar,
                       driver = sp,
                       kinship = kinship[[chr_id]],
                       intcovar = intcovar,
                       method = "double-lod-diff")
ggplotly(autoplot(med2_scan) +
           ggtitle(pheno_name))
```

```{r}
med2_lod <- intermediate::mediator_lod(
  mediator = mrna$expr,
  driver = sp,
  annotation = mrna$annot,
  covar_med = covar,
  kinship = kinship[[chr_id]],
  intcovar = intcovar)
med2_signif <- med2_lod$lod >= 5
# Add info column.
med2_lod$info <- paste("chr =", med2_lod$chr)
```

```{r med2_test,fig.width=9,fig.height=6}
med2_test <- intermediate::mediation_test(
  target   = pheno_data[,pheno_name,drop=FALSE],
  mediator = mrna$expr[, med_signif, drop = FALSE],
  annotation = mrna$annot,
  covar_tar = covar,
  covar_med = covar,
  kinship = kinship[[chr_id]],
  intcovar = intcovar,
  driver = sp)
ggplotly(autoplot(med2_test))
```

```{r}
(sum_med2 <- summary(med2_test))
```

```{r}
if(resultpath != "")
  write.csv(sum_med2, file = paste0(pheno_name, "_mediation_snp.csv"))
```

```{r med2_triad,fig.width=9,fig.height=6}
med2_triad <- intermediate::mediation_triad(
  target   = pheno_data[,pheno_name,drop=FALSE],
  mediator = mrna$expr[, sum_med$id[1], drop = FALSE],
  driver = sp,
  covar_tar = covar,
  kinship = kinship[[chr_id]],
  sdp = ts$sdp,
  label_fn = NULL,
  group_fn = NULL)
autoplot(med2_triad, tname = pheno_name, mname = sum_med$symbol[1], dname = peak_snp)
```

```{r}
if(params$extraSteps != "")
  source(params$extraSteps)
```

