---
params:
  target_name:
    label: Target
    value: GLP1_G83_ins_secrete_gm
  mediator_name:
    label: Mediator
    value: Ptpn18
  chr_id:
    label: Chromosome
    value: 1
  pos_Mbp:
    input: numeric
    value: 90
    label: Peak position
  SNPlevels:
    choices:
      - 2
      - 3
    input: select
    value: 3
    label: Number of SNP levels
  coefType:
    choices:
    - coef
    - blup
    input: select
    label: Coefficient calculation
    value: coef
  dataSetup:
    choices:
    - dataJax.R
    - dataMadison.R
    - dataJaxMadison.R
    input: select
    label: Data setup file
    value: dataJaxMadison.R
  docTitle:
    label: Document title
    value: Mediation Scan
  snpScan:
    input: numeric
    label: Width for SNP scan (0 = whole chromosome)
    value: 10
  offset:
    label: Offset width for fine mapping
    value: 5
  drivers:
    label: Dual or lone drivers
    choices:
    - dual
    - lone
    value: dual
  resultpath:
    input: text
    label: Path to put CSV results (none if '')
    value: '.'
  docAuthor:
    label: Author
    value: '[Brian S Yandell](brian.yandell@wisc.edu)'
  docOutput:
    choices:
    - html_document
    - pdf_document
    - word_document
    input: select
    label: Output format
    value: html_document
  datapath:
    input: text
    label: Path to derived data (required for Madison)
    value: ../data
  showPeaks:
    label: Show highest peaks (0=no, n=number to show)
    value: 0
  extraSteps:
    label: Exra steps file (optional)
    value: ''
---

---
title: "`r paste(params$docTitle, params$mediator_name, params$target_name)`"
author: "`r params$docAuthor`"
date: "`r  format(Sys.time(), '%d %B %Y')`"
output:
  "`r params$docOutput`":
    toc: true
    toc_depth: 3
    code_folding: hide
---

```{r setup, include=FALSE, message = FALSE}
knitr::opts_chunk$set(echo = FALSE, comment = "")
library(dplyr)
library(ggplot2)
library(plotly)
```

```{r}
target_name <- params$target_name
chr_id <- as.character(params$chr_id)
mediator_name <- params$mediator_name
datapath <- params$datapath
resultpath <- params$resultpath
```

```{r}
dataSetup <- file.path("../R", params$dataSetup)
```


This document focuses on a `target` (`r target_name`), a chromosome (`r chr_id`), and a `mediator` (`r mediator_name`). There are three analyses, using drivers based on 1) allele genotype probabilities, 2) SNP genotype probabilities, and 3) SNP pattern to impute genotype probabilities. The driver is based on the best SNP distribution pattern for the `target` but is chosen at the peak LOD for the SNP with that distribution pattern. Finally, The region within 1.5 LOD (or equivalently in terms of IC) of the peak is retained as the best region.


```{r eval=FALSE}
data.frame(params_value = unlist(params))
```

```{r}
source(dataSetup)
```

```{r}
target <- as.matrix(pheno_data[, target_name, drop = FALSE])
rm(pheno_data)
```

```{r}
addcovar <- qtl2pattern::covar_df_mx(covar)
```

```{r}
map <- query_probs(chr_id)
genoprobs <- map$probs
map <- map$map
kinship <- kinship[[chr_id]]
```

### Peak position for target

```{r}
peaks <- readxl::read_excel(file.path(datapath,"Ex vivo peak summary with QTL count[1].xlsx"),
                            col_types = c(rep("text", 2), rep("numeric", 12), "text")) %>%
  dplyr::filter(lodcolumn == target_name,
                chr == chr_id)
```

```{r ins_peaks}
if(!nrow(peaks)) {
  cat("no peaks for", target_name, "\n")
#  knitr::knit_exit()
}
pos_Mbp <- peaks$pos[1]
if(is.na(pos_Mbp)) {
  pos_Mbp <- params$pos_Mbp
}
cat("using peak position", pos_Mbp, "\n")  
```

```{r peak_mar,message=FALSE}
peak_mar <- qtl2::find_marker(map, chr_id, pos_Mbp)
ap <- qtl2::pull_genoprobpos(genoprobs, peak_mar)
```

### Mediator information

```{r}
mrna <- query_mrna(chr_id)
```

```{r}
mrna_id <- 
  (mrna$annot %>%
     filter(symbol == mediator_name))$id
mrna_pos <- 
  (mrna$peaks %>%
     filter(gene_id == mrna_id, 
            qtl_chr == chr_id))$qtl_pos
```

```{r}
mrna.expr <- as.matrix(mrna$expr[, mrna_id, drop = FALSE])
mrna.annot <- 
  dplyr::filter(
    mrna$annot,
    id == mrna_id)
rm(mrna)
```

## Genome allele scan on chr `r chr_id`

```{r}
m <- qtl2::get_common_ids(genoprobs,
                          target,
                          addcovar,
                          mrna.expr,
                          complete.cases = TRUE)
```

```{r}
# Reduce to common IDs
genoprobs <- subset(genoprobs, ind = m)
target <- target[m,, drop = FALSE]
mediator <- mrna.expr[m,, drop = FALSE]
kinship <- kinship[m,m]
addcovar <- addcovar[m,, drop=FALSE]
```


```{r fig.width=12}
qtl_tar <-
  qtl2::scan1(
    genoprobs = genoprobs,
    pheno = target, 
    kinship = kinship, addcovar = addcovar, cores = 2)
```

```{r}
snpScan <- as.numeric(params$snpScan)
if(is.na(snpScan) || snpScan <= 0) {
  start <- end <- NULL
} else {
  start <- pos_Mbp - snpScan
  end <- pos_Mbp + snpScan
}
```

The mediator `r mediator_name` has a peak nearby at `r mrna_pos`.

```{r fig.width=12}
qtl_med <-
  qtl2::scan1(
    genoprobs = genoprobs,
    pheno = mediator, 
    kinship = kinship, addcovar = addcovar, cores = 2)
```

It is useful to look at the genome scan of target adjusted by the mediator, to see if most of the signal is removed.

```{r fig.width=12}
qtl_tar_med <-
  qtl2::scan1(
    genoprobs = genoprobs,
    pheno = target, 
    kinship = kinship, addcovar = cbind(addcovar, mediator), cores = 2)
```

```{r fig.width=12}
qtl_med_tar <-
  qtl2::scan1(
    genoprobs = genoprobs,
    pheno = mediator, 
    kinship = kinship, addcovar = cbind(addcovar, target), cores = 2)
```

```{r fig.width=12}
tmp <- cbind(qtl_tar, qtl_med, qtl_tar_med)
colnames(tmp) <- c("target","mediator","adjusted")
autoplot(tmp, map, lodcolumn = 1:3, main = paste(target_name, "given", mediator_name)) +
  geom_vline(xintercept = c(pos_Mbp, mrna_pos), col = "grey")
```

```{r}
# Relationship $T\vert M$
t.m <- qtl2::fit1(cbind(1, as.matrix(mediator)),
                  target,
                  kinship,
                  addcovar)
nind <- length(t.m$ind_lod)
t.m <- t.m$lod
```

```{r}
# Relationship c
m.t <- qtl2::fit1(cbind(1, as.matrix(target)),
                  mediator,
                  kinship,
                  addcovar)$lod
```

## Causal model genome scans

We now combine the above three genome scans with relationships fits of $T\vert M$ and $T\vert M$ to construct the four causal models (causal, reactive, independent and undecided).

```{r}
nmar <- ncol(genoprobs[[chr_id]])
penalty <- c(nmar, nmar, 2*(nmar-1), 2*nmar-1,0,0) * log10(nind) / 2
```

```{r}
qtl_cmst <- cbind(qtl_med, qtl_tar, qtl_tar, qtl_tar_med, qtl_tar, qtl_med_tar)
colnames(qtl_cmst) <- c("causal","reactive","independent","undecided","mediation","joint")
qtl_cmst[,1] <- t.m + qtl_med[,1]
qtl_cmst[,2] <- m.t + qtl_tar[,1]
qtl_cmst[,3] <- qtl_tar[,1] + qtl_med[,1] 
qtl_cmst[,4] <- qtl_tar_med[,1] + t.m + qtl_med[,1] 
qtl_cmst[,5] <- qtl_tar[,1] - qtl_tar_med[,1]
qtl_cmst[,6] <- qtl_med_tar[,1] + m.t + qtl_tar[,1]
```

```{r}
rm(qtl_tar, qtl_med, qtl_tar_med)
```

```{r eval=FALSE, fig.width=12}
autoplot(qtl_cmst, map, lodcolumn = c(4,6),
         main = paste(target_name, "given", mediator_name)) +
  geom_vline(xintercept = c(mrna_pos, pos_Mbp), col = "grey")
```

```{r fig.width=12}
autoplot(qtl_cmst, map, lodcolumn = 1:4,
         main = paste(target_name, "given", mediator_name)) +
  geom_vline(xintercept = c(mrna_pos, pos_Mbp), col = "grey")
```

Here we focus in on the interesting region. From this it looks like the undecided model is best, but this is simply an artefact of having a richer model (more parameters). 

```{r}
qtl2::find_peaks(qtl_cmst[,1:4],
                 map) %>%
  arrange(desc(lod))
```

```{r fig.width=12}
autoplot(qtl_cmst, map, lodcolumn = 1:4, xlim = c(start,end),
         main = paste(target_name, "given", mediator_name)) +
  geom_vline(xintercept = c(mrna_pos, pos_Mbp), col = "grey")
```

```{r}
out <- qtl2::find_peaks((qtl_cmst - matrix(penalty, nrow(qtl_cmst), ncol(qtl_cmst), byrow = TRUE))[,1:4],
                 map, threshold = -100) %>%
  arrange(desc(lod))
```

Adjust for model complexity using BIC, then the ``r out$lodcolumn`` model comes out on top.

```{r}
out
```

```{r fig.width=12}
autoplot(qtl_cmst - matrix(penalty, nrow(qtl_cmst), ncol(qtl_cmst), byrow = TRUE),
         map, lodcolumn = 1:4, xlim = c(start, end),
         main = paste(target_name, "given", mediator_name),
         ylab = "BIC on log10 scale") +
  geom_vline(xintercept = c(mrna_pos, pos_Mbp), col = "grey")
```

Below is a graph of the mediation scan, now presented as the difference in LOD for the target and the target adjusted by mediator. Notice that the peak mediation scan is at `r qtl2::find_peaks(qtl_cmst[,5, drop = FALSE], map, threshold = -100)`, which need not be near the `target` peak (`r pos_Mbp`) or `mediator` peak (`r mrna_pos`).

```{r fig.width=12}
autoplot(qtl_cmst, map, lodcolumn = 5,
         xlim = c(min(pos_Mbp, start), max(pos_Mbp, end)),
         main = paste(target_name, "given", mediator_name),
         ylab = "mediation LOD drop") +
  geom_vline(xintercept = c(mrna_pos, pos_Mbp), col = "grey")
```

## Mediation test with allele probabilities at mediator peak

```{r}
mrna.annot$driver <- qtl2::find_marker(map, chr_id, mrna.annot$qtl_pos)
driver_med <- genoprobs[[chr_id]][,,unique(mrna.annot$driver), drop = FALSE]
mrna_driver <- (mrna.annot %>% filter(symbol == mediator_name))$driver
```

```{r}
med_med <- intermediate::mediation_test(target = target,
         mediator = mediator,
         annotation = mrna.annot,
         covar_tar = covar,
         covar_med = covar,
         kinship = kinship,
         driver = genoprobs[[chr_id]][,,mrna_driver])
```

```{r}
(sum_med_med <- summary(med_med) %>%
   select(symbol,chr,pos,pvalue,triad,mediation,LR,everything())) %>%
  select(symbol,chr,pos,pvalue,triad,mediation,LR) %>%
  rename(lod = "LR") %>%
  mutate(lod = lod / log(10),
         mediation = mediation / log(10))
```

Here is a plot of allele effects adjusted by mediator at this peak position.

```{r fig.width=12}
med_triad <- intermediate::mediation_triad(target = target,
         mediator = mediator,
         driver = genoprobs[[chr_id]][,,mrna_driver],
         sdp = NULL,
         covar_tar = addcovar,
         covar_med = addcovar,
         kinship = kinship,
         label_fn = NULL, group_fn = NULL)
med_triad$data$label <- NULL
autoplot(med_triad, fitline = TRUE,
         tname = target_name, mname = params$mediator_name, dname = peak_mar)
```

# Mediation test on SNPs in peak region.

```{r}
if(params$SNPlevels == 3) {
  genoprobs <- query_probs(chr = chr_id, start = start, stop = end, allele = FALSE)
  map <- genoprobs$map
  genoprobs <- genoprobs$probs
} else {
  genoprobs <- query_probs(chr = chr_id, start = start, stop = end)
  map <- genoprobs$map
  genoprobs <- genoprobs$probs
}
```

```{r med_joint,fig.width=12}
med_index <- 
  intermediate::mediation_qtl2(target,
                               mediator,
                               mrna.annot, covar, covar, kinship,
                               genoprobs, map,
                               drop_lod = 1.5, query_variant,
                               cores = 4)
```

```{r}
plot(med_index)
```

```{r}
plot(med_index, "IC")
```

```{r}
write.csv(
  med_index$best %>%
    select(-lod) %>%
    rename(lod = "LR") %>%
    mutate(lod = lod / log(10)),
  file = file.path(
    resultpath, 
    paste0(target_name, "_", chr_id, "_", mediator_name, "_topsnps.csv")))
```

```{r}
knitr::knit_exit()
```

## Mediation triad with allele probs

We look again at the mediation triad plots of data. The first uses the SNP at the target. It overlays unrestricted line fits by SNP level.

**This needs to get driver from med_index fit or somewhere.**

```{r fig.width=12}
med_triad <- intermediate::mediation_triad(target = target,
         mediator = mediator,
         driver = genoprobs[[chr_id]][,,mrna_driver],
         sdp = med_index$best$sdp[1],
         covar_tar = addcovar,
         covar_med = addcovar,
         kinship = kinship)
autoplot(med_triad, tname = target_name, mname = params$mediator_name, dname = peak_mar)
```

This second set of plots uses the best SNP for the causal model.

```{r fig.width=12}

med_snp_triad <- intermediate::mediation_triad(target = target,
         mediator = mediator,
         driver = driver_med[,, med_index$best$driver[1]],
         sdp = med_index$best$sdp[1],
         covar_tar = addcovar,
         covar_med = addcovar,
         kinship = kinship,
         group_fn = NULL, label_fn = NULL)
autoplot(med_snp_triad, fitline = TRUE,
         tname = target_name, mname = params$mediator_name, dname = peak_snps[3]) +
  aes(col = group)
```

# SNP Pattern scan

** This needs rework.**

```{r}
patprobs <- 
  qtl2pattern::genoprob_to_patternprob(
    pairprobs[,chr_id], ts$sdp[1])
```

```{r fig.width=12}
pat_tar <- qtl2::scan1(genoprobs = patprobs, 
                      pheno = target, kinship = kinship,
                      addcovar = addcovar)
pat_med = qtl2::scan1(genoprobs = patprobs,
                      pheno = mediator, kinship = kinship,
                      addcovar = addcovar)
pat_tar_med = qtl2::scan1(genoprobs = patprobs,
                      pheno = target, kinship = kinship,
                      addcovar = cbind(addcovar, mediator))
tmp <- cbind(pat_tar, pat_med, pat_tar_med)
colnames(tmp) <- c("target","mediator","adjusted")
autoplot(tmp, map, lodcolumn = 1:3, main = paste(target_name, "given", mediator_name))
rm(tmp)
```

```{r}
nmar <- ncol(patprobs[[chr_id]])
penalty <- c(nmar, nmar, 2*(nmar-1), 2*nmar-1,0) * log10(nind) / 2
```

```{r}
pat_cmst <- cbind(pat_med, pat_tar, pat_tar, pat_tar_med, pat_tar)
colnames(pat_cmst) <- c("causal","reactive","independent","undecided","mediation")
pat_cmst[,1] <- t.m + pat_med[,1]
pat_cmst[,2] <- m.t + pat_tar[,1]
pat_cmst[,3] <- pat_tar[,1] + pat_med[,1] 
pat_cmst[,4] <- pat_tar_med[,1] + t.m + pat_med[,1] 
pat_cmst[,5] <- pat_tar[,1] - pat_tar_med[,1]
```

```{r}
rm(pat_tar, pat_med, pat_tar_med)
```

```{r fig.width=12}
autoplot(pat_cmst, map, lodcolumn = 1:4,
         main = paste(target_name, "given", mediator_name)) +
  geom_vline(xintercept = c(mrna_pos, pos_Mbp), col = "grey")
```

```{r fig.width=12}
autoplot(pat_cmst - matrix(penalty, nrow(pat_cmst), ncol(pat_cmst), byrow = TRUE),
         map, lodcolumn = 1:4, xlim = c(start, end),
         main = paste(target_name, "given", mediator_name)) +
  geom_vline(xintercept = c(mrna_pos, pos_Mbp), col = "grey")
```

```{r}
mrna.annot$driver <- qtl2::find_marker(map, chr_id, mrna.annot$qtl_pos)
driver_signif <- mrna.annot$driver
m <- which(driver_signif %in% dimnames(patprobs[[chr_id]])[[3]])
driver_signif <- driver_signif[m]
driver_pat <- patprobs[[chr_id]][,,unique(c(peak_mar, driver_signif)), drop = FALSE]
```

```{r}
med_pat <- intermediate::mediation_test(target = target,
         mediator = mediator,
         annotation = mrna.annot,
         covar_tar = covar,
         covar_med = covar,
         kinship = kinship,
         driver = NULL,
         driver_med = driver_pat)
```

```{r}
(sum_med_pat <- summary(med_pat) %>%
   select(symbol,chr,pos,pvalue,triad,mediation,LR,everything())) %>%
  select(symbol,chr,pos,pvalue,triad,mediation,LR) %>%
  rename(lod = "LR") %>%
  mutate(lod = lod / log(10),
         mediation = mediation / log(10))
```

```{r}
med_pat$test %>%
  rename(lod = LR) %>%
  mutate(lod = lod / log(10),
         IC = IC / log(10))
```

