---
params:
  target_name:
    label: Phenotype
    value: GLP1_G83_ins_secrete_gm
  chrID:
    label: Chromosome
    value: 1
  coefType:
    choices:
    - coef
    - blup
    input: select
    label: Coefficient calculation
    value: coef
  dataSetup:
    choices:
    - dataJax.R
    - dataMadison.R
    - dataJaxMadison.R
    input: select
    label: Data setup file
    value: dataJaxMadison.R
  docTitle:
    label: Document title
    value: Mediation Scan
  snpScan:
    input: numeric
    label: Width for SNP scan (0 = whole chromosome)
    value: 10
  offset:
    label: Offset width for fine mapping
    value: 2
  resultpath:
    input: text
    label: Path to put CSV results (none if '')
    value: '.'
  docAuthor:
    label: Author
    value: '[Brian S Yandell](brian.yandell@wisc.edu)'
  docOutput:
    choices:
    - html_document
    - pdf_document
    - word_document
    input: select
    label: Output format
    value: html_document
  datapath:
    input: text
    label: Path to derived data (required for Madison)
    value: ..
  showPeaks:
    label: Show highest peaks (0=no, n=number to show)
    value: 0
  useIntcov:
    choices:
    - no
    - yes
    label: Use interacting covariate (sex)?
    value: no
  extraSteps:
    label: Exra steps file (optional)
    value: ''
---

---
title: "`r params$docTitle`"
author: "`r params$docAuthor`"
date: "`r  format(Sys.time(), '%d %B %Y')`"
output:
  html_document:
    toc: true
    toc_depth: 3
    code_folding: hide
---

```{r setup, include=FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(ggplot2)
library(plotly)
```

```{r}
target_name <- params$target_name
chr_id <- as.character(params$chrID)
datapath <- params$datapath
resultpath <- params$resultpath
```

This is built off script developed by Dan Gatti. In Rstudio, use `Knit with Parameters` from <kbd>Knit</kbd> pulldown to adjust parameter settings. See [README.md](README.md) for details.

This document has a variety of strategies, but here is current recommendation. Assumed is that we have identified a `target` (`r target_name`) and a chromosome (`r chr_id`), and that we have information on peaks for the `target` and potential `mediators`.

- scan for best `mediator` of `target`
    + use `target` whole genome (allele-based) scan to identify peak
    + identify `mediators` that have peak within region (`r params$snpScan` Mbp of peak)
    + use allele probabilities at mediator peaks as `drivers`
    + use 3-level SNP probabilites at mediator SNP peaks as `drivers`
    + maybe use 2-driver model but do not simply rely on 1 driver at pheno peak
- scan for best driver of `target` and `mediator`
    + use region and 3-level SNP probabilities (as `drivers`) already identified
    + for each `driver`, compute joint LOD (`undecided` model) of `target` and `mediator`
    + identify sub-region of highest joint LOD (within 1.5 of peak)
    + perform CMST on all drivers within sub-region
    + restrict to drivers that have highest recomputed joint LOD within sub-region
    + find best `pvalue`s and `triad`, noting SNP distribution pattern (`sdp`)

The reason to do multiple checks is the following:

* allele-based is smoother but cruder: useful for identifying broad region
* joint LOD identifies most likely region for best fit
* pattern of missing data important to balance in CMST tests, changing joint LOD
* need to consider relative magnitude of joint LOD, `sdp` and pvalue together
    + can get more significant `pvalue` for other `sdp`s that are worse models

```{r}
data.frame(params_value = unlist(params))
```

### Data Setup

```{r}
source(params$dataSetup)
kinship <- kinship[[chr_id]]
```

```{r}
target <- pheno_data[, target_name, drop = FALSE]
```

```{r}
addcovar <- qtl2pattern::covar_df_mx(covar)
```

```{r}
intcovar <- NULL
```

### Genome scan information

```{r}
genoprobs <- query_probs()$probs
map <- query_probs()$map
```

```{r}
(peaks <- readxl::read_excel("local/Jax/Ex vivo peak summary with QTL count[1].xlsx",
                            col_types = c(rep("text", 2), rep("numeric", 12), "text")) %>%
  dplyr::filter(lodcolumn == target_name,
                chr == chr_id))
```

```{r}
if(!nrow(peaks)) {
  cat("no peaks for", target_name, "\n")
  knitr::knit_exit()
}
```

```{r ins_peaks}
pos_Mbp <- peaks$pos[1]
```

```{r peak_mar,message=FALSE}
peak_mar <- qtl2::find_marker(map, chr_id, pos_Mbp)
ap <- qtl2::pull_genoprobpos(genoprobs, peak_mar)
```

### mRNA information

```{r}
mrna <- query_mrna(chr_id)
```

Find peaks for local mRNA that are close to SNP peak.

```{r}
med_signif <- 
  (dplyr::filter(
    mrna$peaks,
    qtl_chr == chr_id,
    abs(qtl_pos - pos_Mbp) <= params$snpScan,
    gene_start >= pos_Mbp - params$snpScan,
    gene_end <= pos_Mbp + params$snpScan))$gene_id
```

```{r}
mrna.expr <- mrna$expr[, med_signif, drop = FALSE]
mrna.annot <- 
  dplyr::filter(
    mrna$annot,
    id %in% med_signif)
rm(mrna)
```

### Mediation with allele probs.

```{r}
mrna.annot$driver <- qtl2::find_marker(map, chr_id, mrna.annot$qtl_pos)
driver_med <- genoprobs[[chr_id]][,,unique(mrna.annot$driver), drop = FALSE]
```

```{r med_test,fig.width=9,fig.height=6}
med_test <- intermediate::mediation_test(
  target   = target,
  mediator = mrna.expr,
  annotation = mrna.annot,
  covar_tar = covar,
  covar_med = covar,
  kinship = kinship,
  intcovar = intcovar,
  driver = NULL,
  driver_med = driver_med)
autoplot(med_test)
```

```{r}
(sum_med <- 
   dplyr::arrange(
     summary(med_test),
     pvalue))
```

```{r}
if(resultpath != "")
  write.csv(sum_med, file = file.path(params$resultpath,
                                      paste0(target_name, "_", chr_id, "_mediation.csv")))
```

#### Driver effects by Mediator, Target, Adjusted model.

```{r}
ggplot(med_test$best %>%
         filter(pvalue <= 0.05) %>%
         mutate(mediation = mediation / log(10))) +
  aes(mediation, -log10(pvalue), col = biotype) +
  geom_point()
```

```{r}
med_effects <- intermediate::mediation_effect(med_test)
autoplot(med_effects)
```

### Association mapping

Map all of Chr `r chr_id` unless `snpScan` (`r params$snpScan`) is positive.

```{r}
snpScan <- as.numeric(params$snpScan)
if(is.na(snpScan) || snpScan <= 0) {
  start <- end <- NULL
} else {
  start <- pos_Mbp - snpScan
  end <- pos_Mbp + snpScan
}
```

```{r}
pairprobs <- query_probs(chr = chr_id, start = start, stop = end, allele = FALSE)
pairmap <- pairprobs$map
pairprobs <- pairprobs$probs
```

```{r ins_assoc_chr,fig.width=12}
if(is.null(start) & is.null(end)) {
  cat(paste("scan1snps of chromosome", chr_id, "\n"), file = stderr())
} else {
  cat(paste("scan1snps of chromosome", chr_id, "from", start, "to", end, "\n"), file = stderr())
}
  
assoc_ins = qtl2::scan1snps(genoprobs = pairprobs[,chr_id], map = pairmap, 
                      pheno = target, kinship = kinship,
                      addcovar = addcovar, intcovar = intcovar, chr = chr_id, start = start,
                      end = end, query_func = query_variant, cores = 4,
                      keep_all_snps = FALSE)
autoplot(assoc_ins$lod, assoc_ins$snpinfo, show_all_snps = FALSE)
```

```{r ins_assoc_rescan,fig.width=12}
(ts <- qtl2::top_snps(assoc_ins$lod, assoc_ins$snpinfo) %>%
  arrange(desc(lod))) %>%
  head(10)
```

```{r}
if(resultpath != "")
  write.csv(ts, file = file.path(params$resultpath,
                                 paste0(target_name, "_", chr_id, "_topsnps.csv")))
```

View patterns in top SNPs.

```{r assoc_ins_pattern,fig.width=9,fig.height=6}
autoplot(assoc_ins$lod, assoc_ins$snpinfo, patterns = "all", drop_hilit = 5, main = target_name)
```

### Top missense, splice or stop SNPs

```{r ins_top_snps_left}
(ts <- left_join(ts %>% dplyr::rename(id = "ensembl_gene"),
          mrna.annot %>% select(id, symbol),
          by = "id")) %>%
  select(snp_id, lod, pos, id, symbol, consequence) %>%
  head(10)
```

```{r}
if(resultpath != "")
  write.csv(ts, 
          file = file.path(params$resultpath, 
                           paste0(target_name, "_", chr_id, "_topsnps_peak.csv")))
```

```{r}
ts %>% dplyr::filter(grepl("splice|missense|stop", consequence))
```

### Mediation with SNP probs

Find best SNP for each mediator.

```{r}
assoc_med = qtl2::scan1snps(genoprobs = pairprobs[,chr_id], map = pairmap, 
                      pheno = mrna.expr,
                      kinship = kinship,
                      addcovar = addcovar, intcovar = intcovar, chr = chr_id, start = start,
                      end = end, query_func = query_variant, cores = 4,
                      keep_all_snps = FALSE)
```

```{r}
(ts_med <- qtl2pattern::top_snps_all(assoc_med$lod, assoc_med$snpinfo) %>%
  arrange(desc(lod))) %>%
  head(10)
```

```{r}
(ts_med <- ts_med %>%
   group_by(pheno, chr) %>%
   summarize(
     pos = pos[which.max(lod)],
     snp_id = snp_id[which.max(lod)],
     lod = max(lod)) %>%
   ungroup %>%
   arrange(desc(lod)))
```

Set up drivers for mediators.

```{r}
peak_snp <- ts$snp_id[1]
```

```{r}
m <- match(mrna.annot$id, ts_med$pheno)
mrna.annot$driver <- ts_med$snp_id[m]
driver_med_snp <-
  qtl2::genoprob_to_snpprob(
    pairprobs, 
    assoc_ins$snpinfo)[[chr_id]][,, c( peak_snp, unique(ts_med$snp_id)), drop = FALSE]
```

Mediation scan using best SNP for target `r target_name`. Will generally not be best for mediators.
(Want to change this to using mediator drivers as in mediateOne.Rmd. We use a hidden feature of `mediation_scan`.)

```{r med2_scan,fig.width=9,fig.height=6}
med2_scan <- intermediate::mediation_scan(target  = target,
                       mediator = mrna.expr,
                       annotation = mrna.annot,
                       covar    = addcovar,
                       driver = driver_med_snp,
                       kinship = kinship,
                       intcovar = intcovar,
                       method = "double-lod-diff")
ggplotly(autoplot(med2_scan) +
           ggtitle(target_name))
```

```{r medd_scan,fig.width=9,fig.height=6}
medd_scan <- intermediate::mediation_scan(target  = target,
                       mediator = mrna.expr,
                       annotation = mrna.annot,
                       covar    = addcovar,
                       driver = driver_med_snp[,,1],
                       kinship = kinship,
                       intcovar = intcovar,
                       method = "double-lod-diff")
ggplotly(autoplot(medd_scan) +
           ggtitle(target_name))
```

```{r med2_test,fig.width=9,fig.height=6}
med2_test <- intermediate::mediation_test(
  target   = target,
  mediator = mrna.expr,
  annotation = mrna.annot,
  covar_tar = covar,
  covar_med = covar,
  kinship = kinship,
  intcovar = intcovar,
  driver = NULL,
  driver_med = driver_med_snp)
autoplot(med2_test)
```

```{r}
(sum_med2 <- summary(med2_test))
```

```{r}
if(resultpath != "")
  write.csv(sum_med2, file = file.path(params$resultpath,
                                       paste0(target_name, "_", chr_id, "_mediation_snp.csv")))
```

```{r med2_triad,fig.width=9,fig.height=6}
m <- match(sum_med$id[1], ts_med$pheno)
med2_triad <- intermediate::mediation_triad(
  target   = target,
  mediator = mrna.expr[, sum_med$id[1], drop = FALSE],
  driver = driver_med_snp[,, ts_med$snp_id[m]],
  covar_tar = covar,
  kinship = kinship,
  sdp = ts$sdp,
  label_fn = NULL,
  group_fn = NULL)
autoplot(med2_triad, tname = target_name, mname = sum_med$symbol[1], dname = peak_snp)
```

```{r}
if(params$extraSteps != "")
  source(params$extraSteps)
```

Driver effects by Mediator, Target, Adjusted model.

```{r}
ggplot(med2_test$best %>%
         filter(pvalue <= 0.05) %>%
         mutate(mediation = mediation / log(10))) +
  aes(mediation, -log10(pvalue), col = biotype) +
  geom_point()
```

```{r}
med_effects <- intermediate::mediation_effect(med2_test)
autoplot(med_effects)
```

